src/DistIL/IR/MatchExtensions.cs

```diff
@@ -0,0 +1,254 @@
+namespace DistIL.IR;
+
+using DSL.PatternArguments;
+using DSL;
+using Utils.Parser;
+
+public static class MatchExtensions
+{
+    public static bool Match(this Instruction instruction, string pattern, out OutputPattern outputs)
+    {
+        outputs = new OutputPattern(pattern);
+
+        return MatchInstruction(instruction, outputs.Pattern!, outputs);
+    }
+
+    public static bool Match(this Instruction instruction, string pattern)
+    {
+        var outputs = new OutputPattern(pattern);
+
+        return MatchInstruction(instruction, outputs.Pattern!, outputs);
+    }
+
+    internal static bool MatchInstruction(this Instruction instruction, InstructionPattern instrPattern, OutputPattern outputs)
+    {
+        if (instruction is BinaryInst bin) {
+            return MatchBinary(bin, instrPattern, outputs);
+        }
+        if (instruction is CompareInst comp) {
+            return MatchCompare(comp, instrPattern, outputs);
+        }
+        else if (instruction is UnaryInst un) {
+            return MatchUnary(un, instrPattern, outputs);
+        }
+
+        return MatchOtherInstruction(instruction, instrPattern, outputs);
+    }
+
+    private static bool MatchOtherInstruction(Instruction instruction, InstructionPattern pattern, OutputPattern outputs)
+    {
+        var op = pattern.OpCode;
+        var ops = MatchOpCode((op, instruction));
+
+        return MatchOperands(instruction, pattern, outputs);
+    }
+
+    private static bool MatchOpCode((Opcode op, Instruction instruction) opInstrTuple)
+    {
+        return opInstrTuple switch {
+            (Opcode.Unknown, _) => false,
+            (Opcode.Goto, BranchInst) => true,
+            (Opcode.Switch, SwitchInst) => true,
+            (Opcode.Ret, ReturnInst) => true,
+            (Opcode.Phi, PhiInst) => true,
+            (Opcode.Select, SelectInst) => true,
+            (Opcode.Lea, PtrOffsetInst) => true,
+            (Opcode.Getfld,FieldExtractInst) => true,
+            (Opcode.Setfld, FieldInsertInst) => true,
+            (Opcode.ArrAddr, ArrayAddrInst) => true,
+            (Opcode.FldAddr, FieldAddrInst) => true,
+            (Opcode.Load, LoadInst) => true,
+            (Opcode.Store, StoreInst) => true,
+            (Opcode.Conv, ConvertInst) => true,
+            _ => false
+        };
+    }
+
+    private static bool MatchOperands(Instruction instruction, InstructionPattern pattern, OutputPattern outputs)
+    {
+        bool matched = true;
+
+        if (pattern.Arguments.Count > instruction.Operands.Length) {
+            return false;
+        }
+
+        for (int index = 0; index < pattern.Arguments.Count; index++) {
+            Value? operand = instruction.Operands[index];
+            matched &= MatchValue(operand, pattern.Arguments[index], outputs);
```

Comment:
nit: matching can stop as soon as a false match is found (not super important, but maybe useful until we get to do the source generator thing)